### PR TITLE
test(examples): e2e test for examples

### PR DESCRIPTION
**Pull request by akihikokuroda**
### Which issue(s) does this pull-request address?

Closes: #62 

### Description

This PR changes add making examples run during e2e testing.

### Checklist

- [x] I have read the [contributor guide](https://github.com/i-am-bee/bee-agent-framework/blob/main/CONTRIBUTING.md)
- [x] Linting passes: `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: `yarn format` or `yarn format:fix`
- [x] Unit tests pass: `yarn test:unit`
- [ ] E2E tests pass: `yarn test:e2e`
- [ ] Tests are included 
- [ ] Documentation is changed or added
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

### CODE REVIEW

**Comment by akihikokuroda** [comment]
For now, there is only one test for "memory/agentMemory.ts" added.  I want to know if this is in the right direction or not.
Thanks!

**Review by Tomas2D — CHANGES_REQUESTED** [review]
- We don't want to modify the existing `examples` directory.
- Testing output is not important.

Please see the Python example that I mentioned in the related Github issue.

**Comment by akihikokuroda** [comment]
@Tomas2D There are not changes in the `/examples` directory.  The tests of the examples are under `/tests/examples` directory that executes the examples as-is and check if the examples generate any output.

**Comment by Tomas2D** [comment]
Oh. I wanted to say that I don't want 1:1 mapping between the `tests/examples` and the `examples` directories.

I would like to have one example test file, `test/e2e/examples.ts`, which executes (recursively) all files under the `examples` directory. This file will contain filters (so you can ignore some concrete examples or run them conditionally).

File: .github/workflows/examples-tests.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+jobs:
+  test:
+    name: Tests
+    timeout-minutes: 40
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
+      matrix:
+        node-version: [22.x]
```
Comment (Tomas2D) [inline .github/workflows/examples-tests.yml:23]: run with 18.x

File: .github/workflows/examples-tests.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
+name: example Tests
```
Comment (Tomas2D) [inline .github/workflows/examples-tests.yml:1]: E2E Exampes

**Comment by akihikokuroda** [comment]
@Tomas2D Thanks for review and direction.  I'll work on it.

**Comment by akihikokuroda** [comment]
`yarn test:examples` runs the all examples
`EXAMPLE_NAME=<example name pattern>  yarn test:examples` runs the tests that matches the pattern

The interactive examples and some other are failing.  I'll look into the failures.

File: tests/e2e/examples.test.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+for (const example of files) {
+  if (pattern == null || pattern.trim() == "") {
+    runExample(example);
+  } else {
+    if (example.includes(pattern)) {
+      runExample(example);
+    }
+  }
+}
```
Comment (Tomas2D) [inline tests/e2e/examples.test.ts:57]: The script could be clearer and easier to follow. However, you got the point.

Let's just simply it to something like this?

```bash
# glob is already installed (just types are missing)
yarn add -D @types/glob
```

```typescript
import { exec } from "child_process";
import { promisify } from "util";
import { glob } from "glob";

const execAsync = promisify(exec);
const pattern = process.env.EXAMPLE_NAME || `./examples/**/*.ts`;

describe("e2e test", async () => {
  const exampleFiles = await glob(pattern, {
    cwd: process.cwd(),
    dot: false,
    realpath: true,
  });

  it.each(exampleFiles)("Runs %s", async (example) => {
    const { stderr, stdout } = await execAsync(`yarn start -- ${example}`);
    console.log({
      path: example,
      stdout,
    });
    expect(stderr).toBeFalsy();
  });
});
```


Also, please try to play with the Vitest concurrency. We will not typically want to run all examples in a sequence but rather in parallel (with some limit).

**Review by Tomas2D — CHANGES_REQUESTED** [review]

File: tests/e2e/examples.test.ts (hunk 3 above)
Comment (akihikokuroda) [reply tests/e2e/examples.test.ts]: Thanks for the updates.  It is much simpler now.  I changed to run the examples in parallel.  I haven't add the `some limit` yet.

File: .github/workflows/examples-tests.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      - name: Run ollama
+        run: |
+          ollama serve &
+          ollama pull llama3.1
+      - name: Call ollama API
+        run: |
+          curl -d '{"model": "llama3.1:latest", "stream": false, "prompt":"Whatever I say, asnwer with Yes"}' http://localhost:11434/api/generate
+      - name: Example Tests
+        run: |
+          yarn test:examples
```
Comment (Tomas2D) [inline .github/workflows/examples-tests.yml:47]: You'll need to set `secrets` to run examples.
See: https://github.com/i-am-bee/bee-agent-framework/blob/main/.github/workflows/e2e-tests.yml#L36-L45

File: .github/workflows/examples-tests.yml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        run: yarn install --immutable
+      - name: Install ollama
+        run: curl -fsSL https://ollama.com/install.sh | sh
+      - name: Run ollama
+        run: |
+          ollama serve &
+          ollama pull llama3.1
+      - name: Call ollama API
+        run: |
+          curl -d '{"model": "llama3.1:latest", "stream": false, "prompt":"Whatever I say, asnwer with Yes"}' http://localhost:11434/api/generate
```
Comment (Tomas2D) [inline .github/workflows/examples-tests.yml:44]: I am a little bit worried that takes forever. But let's give it a shot.

File: tests/setup.examples.ts
Comment (Tomas2D) [inline tests/setup.examples.ts:1]: Why is this file needed?

File: vitest.examples.config.ts
Comment (Tomas2D) [inline vitest.examples.config.ts:1]: Why is this file needed?

**Review by Tomas2D — CHANGES_REQUESTED** [review]

File: tests/setup.examples.ts
Comment (akihikokuroda) [reply tests/setup.examples.ts:1]: I took out `onTestFailed`.
Screenshot 2024-10-11 at 4 08 20 PM

File: .github/workflows/examples-tests.yml (hunk 5 above)
Comment (akihikokuroda) [reply .github/workflows/examples-tests.yml:44]: I tried it in my personal repo and it was not too bad.

File: package.json [hunk 6]
```diff
@@ -77,6 +77,8 @@
     "test:e2e:watch": "vitest watch tests",
     "test:all": "vitest run",
     "test:watch": "vitest watch",
+    "test:examples": "vitest --config ./vitest.examples.config.ts run tests/e2e/examples",
+    "test:examples:watch": "vitest --config ./vitest.examples.config.ts run tests/e2e/examples",
```
Comment (Tomas2D) [inline package.json:81]: We must exclude `examples` from standalone `e2e`; otherwise, we would run them twice.

Comment (akihikokuroda) [reply package.json:81]: Is it ok to move the `examples.test.ts` to another directory like `/tests/examples`?

File: tests/e2e/examples.test.ts (hunk 3 above)
Comment (Tomas2D) [reply tests/e2e/examples.test.ts]: Yes. However, in that case, I would not call these `E2E Tests` but rather `Examples Tests` or something like that.

File: package.json (hunk 6 above)
Comment (Tomas2D) [reply package.json:81]: Yes

**Comment by akihikokuroda** [comment]
In the last commit:

- move the examples.test.ts to /tests/examples
- limit the number of concurrent executions to 5

**Comment by akihikokuroda** [comment]
The last commit adds the configuration that can exclude examples from the tests in case some examples may not be expected work successfully.

File: tests/examples/examples.test.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+const pattern = process.env.EXAMPLE_NAME || `./examples/**/*.ts`;
+
+describe("E2E Examples", async () => {
+  const semaphore = new Semaphore(5);
+  const exampleFiles = await glob(pattern, {
+    cwd: process.cwd(),
+    dot: false,
+    realpath: true,
+  });
+  const config = JSON.parse(fs.readFileSync("./tests/examples/example_test.cfg", "utf8"));
```
Comment (Tomas2D) [inline tests/examples/examples.test.ts:31]: I don't see the point of adding a next file. Declare ignored paths in this file directly. 
Note: don't use a sync operation in an async environment

File: tests/examples/examples.test.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+  for (const example of exampleFiles) {
+    if (config.exclude.includes(example)) {
+      continue;
+    }
+    it.concurrent(`Run ${example}`, async () => {
+      await semaphore.acquire();
```
Comment (Tomas2D) [inline tests/examples/examples.test.ts:38]: Why can't you just use https://vitest.dev/config/#sequence-concurrent with https://vitest.dev/config/#maxconcurrency? It seems better.

File: tests/examples/examples.test.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      try {
+        execSync(`yarn start -- ${example}`).toString();
```
Comment (Tomas2D) [inline tests/examples/examples.test.ts:40]: Why do you call `toString()` even though you don't print anything?

Comment (Tomas2D) [reply tests/examples/examples.test.ts:40]: btw, I think you should print something.

File: vitest.examples.config.ts
Comment (Tomas2D) [inline vitest.examples.config.ts:1]: Move this file under `examples`.

File: vitest.examples.config.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import tsConfigPaths from "vite-tsconfig-paths";
+import packageJson from "./package.json";
+
+export default defineConfig({
+  test: {
+    globals: true,
+    passWithNoTests: true,
+    testTimeout: 120 * 1000,
+    printConsoleTrace: true,
+    setupFiles: ["./tests/setup.examples.ts"],
```
Comment (Tomas2D) [inline vitest.examples.config.ts:11]: Use the same setup as the E2E tests.

File: vitest.examples.config.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+      interopDefault: false,
+    },
+  },
+  define: {
+    __LIBRARY_VERSION: JSON.stringify(packageJson.version),
+  },
+  plugins: [
+    tsConfigPaths({
+      projects: ["tsconfig.json"],
+    }),
```
Comment (Tomas2D) [inline vitest.examples.config.ts:22]: This seems wrong - you should use the `tsconfig.examples.json`

File: vitest.examples.config.ts (hunk 10 above)
Comment (Tomas2D) [reply vitest.examples.config.ts:11]: or ideally just import `e2e` test config and spread it here

```
export define defineConfig({
   ...e2eConfig,
   // changes
})
```

**Review by Tomas2D — CHANGES_REQUESTED** [review]

File: vitest.examples.config.ts (hunk 11 above)
Comment (akihikokuroda) [reply vitest.examples.config.ts:22]: I seems that the typescript config file for "/tests/examples/examples.test.ts" must be "tsconfig.json".

File: vitest.examples.config.ts (hunk 10 above)
Comment (akihikokuroda) [reply vitest.examples.config.ts:11]: I'm not sure how to do it.  What necessary is undefine `afterEach` function defined in the "setup.ts".

File: package.json [hunk 12]
```diff
@@ -77,6 +77,8 @@
     "test:e2e:watch": "vitest watch tests",
     "test:all": "vitest run",
     "test:watch": "vitest watch",
+    "test:examples": "vitest --max-concurrency=10 --config ./examples/vitest.examples.config.ts run tests/examples",
+    "test:examples:watch": "vitest --max-concurrency=10 --config ./examples/vitest.examples.config.ts run tests/examples",
```
Comment (Tomas2D) [inline package.json:81]: Instead of passing `max-concurrency` as a flag, you can set it directly within the configuration file `vitest.examples.config.ts`

**Review by Tomas2D — CHANGES_REQUESTED** [review]
So far so good!

File: tests/examples/examples.test.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+import { exec } from "child_process";
+import { glob } from "glob";
+import { promisify } from "util";
+
+const execAsync = promisify(exec);
+const pattern = process.env.EXAMPLE_NAME || `./examples/**/*.ts`;
+
+const exclude: string[] = []; // list of examples that are excluded
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply tests/examples/examples.test.ts:24]: How can I address these?  The ENVs for the test(examples.test.ts) are passed into the examples.  So if the necessary ENVs are defined in the github action, it should work.  So I assume that some examples never get the necessary ENVs.  Do you want to update the examples not to fail so that they don't be in the exclude list?

[the remark this answers is not in the document]
Comment (Tomas2D) [reply tests/examples/examples.test.ts:24]: >  So I assume that some examples never get the necessary ENVs.
It would be great if one could run that locally. Only some have all environments set, so it makes sense to ignore those examples that cannot be run. How? You have to manually look at all examples and find out which need which ENVs. I can do that later if you don't want to.

[the remark this answers is not in the document]
Comment (akihikokuroda) [reply tests/examples/examples.test.ts:24]: I'll do that.  I wanted to know how you want to address these.  Thanks!

[the remark this answers is not in the document]
Comment (akihikokuroda) [reply tests/examples/examples.test.ts:24]: I believe that some tests that wait for user input also fail.  Do you want to exclude those, too?

[the remark this answers is not in the document]
Comment (Tomas2D) [reply tests/examples/examples.test.ts:24]: Good point. You can pass the prompt to them like this.

```
yarn start -- ${example} <<< "Hello world!"
```

**Comment by akihikokuroda** [comment]
There are still 3 example errors.  I added them in the exclude list for now.  They are:

Run examples/cache/fileCache.ts 964ms
Run examples/llms/providers/bam_verbose.ts 1718ms
Run examples/llms/providers/bam.ts 1919ms

Here is the errors.
```
 FAIL  tests/examples/examples.test.ts > E2E Examples > Run examples/cache/fileCache.ts
Error: Command failed: yarn start -- examples/cache/fileCache.ts <<< "Hello world"
node:internal/modules/run_main:129
    triggerUncaughtException(
    ^

[SerializerError: Class "UnconstrainedCache" was not found!] {
  isFatal: true,
  isRetryable: false,
  context: {},
  [errors]: []
}

Node.js v20.17.0

 FAIL  tests/examples/examples.test.ts > E2E Examples > Run examples/llms/providers/bam_verbose.ts
Error: Command failed: yarn start -- examples/llms/providers/bam_verbose.ts <<< "Hello world"
node:internal/modules/run_main:129
    triggerUncaughtException(
    ^

[LLMError: LLM has occurred an error!] {
  isFatal: false,
  isRetryable: false,
  context: {},
  [errors]: [
    HttpError: Model not found
        at clientErrorWrapper (file:///Users/akihikokuroda/development/repositories/agenticflow/bee-agent-framework/node_modules/@ibm-generative-ai/node-sdk/dist/chunk-D3OWGAYN.js:321:13)
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
        at async file:///Users/akihikokuroda/development/repositories/agenticflow/bee-agent-framework/node_modules/p-queue-compat/dist/esm/index.js:118:26 {
      error: 'Not Found',
      status_code: 404,
      extensions: {
        code: 'NOT_FOUND',
        state: { model_id: 'meta-llama/llama-3-1-70b-instruct' }
      },
      [cause]: {
        status_code: 404,
        error: 'Not Found',
        message: 'Model not found',
        extensions: {
          code: 'NOT_FOUND',
          state: { model_id: 'meta-llama/llama-3-1-70b-instruct' }
        }
      }
    }
  ]
}

Node.js v20.17.0

 ❯ clientErrorWrapper node_modules/@ibm-generative-ai/node-sdk/dist/chunk-D3OWGAYN.js:321:13

 FAIL  tests/examples/examples.test.ts > E2E Examples > Run examples/llms/providers/bam.ts
Error: Command failed: yarn start -- examples/llms/providers/bam.ts <<< "Hello world"
node:internal/modules/run_main:129
    triggerUncaughtException(
    ^

[LLMError: LLM has occurred an error!] {
  isFatal: false,
  isRetryable: false,
  context: {},
  [errors]: [
    HttpError: Model not found
        at clientErrorWrapper (file:///Users/akihikokuroda/development/repositories/agenticflow/bee-agent-framework/node_modules/@ibm-generative-ai/node-sdk/dist/chunk-D3OWGAYN.js:321:13)
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
        at async file:///Users/akihikokuroda/development/repositories/agenticflow/bee-agent-framework/node_modules/p-queue-compat/dist/esm/index.js:118:26 {
      error: 'Not Found',
      status_code: 404,
      extensions: {
        code: 'NOT_FOUND',
        state: { model_id: 'meta-llama/llama-3-1-70b-instruct' }
      },
      [cause]: {
        status_code: 404,
        error: 'Not Found',
        message: 'Model not found',
        extensions: {
          code: 'NOT_FOUND',
          state: { model_id: 'meta-llama/llama-3-1-70b-instruct' }
        }
      }
    }
  ]
}

Node.js v20.17.0

 ❯ clientErrorWrapper node_modules/@ibm-generative-ai/node-sdk/dist/chunk-D3OWGAYN.js:321:13

```

**Comment by Tomas2D** [comment]
- Regarding BAM (llama 3.1 is accessible only to some users). Let's ignore it for now. I will resolve this internally.
- Change `EXAMPLE_NAME` to `INCLUDE` or `INCLUDE_PATTERN`
- Add `EXCLUDE` or `EXCLUDE_PATTERN` optional env and use `ignore` property in the `glob` function
- What wraps my head around is this not working test `examples/cache/fileCache.ts`. It works on my machine.
   - Try to 1. run the example without `vitest` (`yarn start examples/cache/fileCache.ts` or 2. replace `os.tmpdir` with some different static path.

**Comment by akihikokuroda** [comment]
The environment variable names are now "INCLUDE_PATTERN" and "EXCLUDE_PATTERN".
The `fileCache.ts` works by replacing the os.tmpdir for me.  Thanks!

File: tests/examples/examples.test.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    cwd: process.cwd(),
+    dot: false,
+    realpath: true,
+    ignore: [exclude_pattern],
+  });
+
+  for (const example of exampleFiles) {
+    if (exclude.includes(example.slice(prefixLength))) {
+      continue;
+    }
```
Comment (Tomas2D) [inline tests/examples/examples.test.ts:54]: This can be simplified.

```
   if (!exclude.startsWith('examples') {
      continue;
    }
```

File: tests/examples/examples.test.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    it.concurrent(`Run ${example}`, async () => {
+      const { stdout } = await execAsync(`yarn start -- ${example} <<< "Hello world"`);
```
Comment (Tomas2D) [inline tests/examples/examples.test.ts:56]: we should also check for `stderr` being empty and return code being equal to 0

**Comment by Tomas2D** [comment]
The `os.tmpdir()` should be cross-platform. Which operating system are you using? This should work

**Comment by akihikokuroda** [comment]
It's M1 Mac.  Sonoma 14.5

File: tests/examples/examples.test.ts (hunk 14 above)
Comment (akihikokuroda) [reply tests/examples/examples.test.ts:54]: The condition is to exclude such as when 
examples is `examples/templates/objects.ts`
and exclude contains `templates/objects.ts`.

So the items in exclude list don't need to have `examples/`.

**Comment by akihikokuroda** [comment]
There are discussions about `os.tmpdir()` on Mac OS.

**Review by Tomas2D — APPROVED** [review]
Great. Thank you for your contribution.

Regarding the `os.tmpdir()` - feel free to propose an alternative in another issue.
